pkg/chubaofs/nodeserver.go

```diff
@@ -234,3 +243,148 @@ func nodeGetVolumeStats(_ context.Context, volumePath string) (*csi.NodeGetVolum
+	for i := range pvList.Items {
+		_, exist := nodePVNames[pvList.Items[i].Name]
+		if exist {
+			nodePVs = append(nodePVs, &pvList.Items[i])
+		}
+	}
+
+	return nodePVs, nil
+}
+
+type persistentVolumeWithPods struct {
+	*v1.PersistentVolume
+	pods []*v1.Pod
+}
+
+// getAttachedPVWithPodsOnNode finds all persistent volume objects as well as the related pods in the node.
+func (ns *nodeServer) getAttachedPVWithPodsOnNode(nodeName string) ([]*persistentVolumeWithPods, error) {
+	pvs, err := ns.getAttachedPVOnNode(nodeName)
+	if err != nil {
+		return nil, fmt.Errorf("getAttachedPVOnNode faied: %v", err)
+	}
+
+	claimedPVWithPods := make(map[string]*persistentVolumeWithPods, len(pvs))
+	for _, pv := range pvs {
+		if pv.Spec.ClaimRef == nil {
+			continue
+		}
+
+		pvcKey := fmt.Sprintf("%s/%s", pv.Spec.ClaimRef.Namespace, pv.Spec.ClaimRef.Name)
+		claimedPVWithPods[pvcKey] = &persistentVolumeWithPods{
+			PersistentVolume: pv,
+		}
+	}
+
+	allPodsOnNode, err := ns.Driver.ClientSet.CoreV1().Pods("").List(metav1.ListOptions{
+		FieldSelector: "spec.nodeName=" + nodeName,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("list pods failed: %v", err)
+	}
+
+	for i := range allPodsOnNode.Items {
+		pod := allPodsOnNode.Items[i]
+
+		for _, volume := range pod.Spec.Volumes {
+			if volume.PersistentVolumeClaim == nil {
+				continue
+			}
+			pvcKey := fmt.Sprintf("%s/%s", pod.Namespace, volume.PersistentVolumeClaim.ClaimName)
+			pvWithPods, ok := claimedPVWithPods[pvcKey]
+			if !ok {
+				continue
+			}
+
+			pvWithPods.pods = append(pvWithPods.pods, &pod)
+		}
+	}
+
+	ret := make([]*persistentVolumeWithPods, 0, len(claimedPVWithPods))
+	for _, v := range claimedPVWithPods {
+		if len(v.pods) != 0 {
+			ret = append(ret, v)
+		}
+	}
+
+	return ret, nil
+}
+
+// remountDamagedVolumes try to remount all the volumes damaged during csi-node restart,
+// includes the GlobalMount per pv and BindMount per pod.
+func (ns *nodeServer) remountDamagedVolumes(nodeName string) {
+	startTime := time.Now()
+
+	pvWithPods, err := ns.getAttachedPVWithPodsOnNode(nodeName)
+	if err != nil {
+		glog.Warningf("get attached pv with pods info failed: %v\n", err)
+		return
+	}
+
+	if len(pvWithPods) == 0 {
+		return
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(len(pvWithPods))
+	for _, pvp := range pvWithPods {
+		go func(p *persistentVolumeWithPods) {
+			defer wg.Done()
+
+			// remount globalmount
+			mountPath := fmt.Sprintf("/var/lib/kubelet/plugins/kubernetes.io/csi/pv/%s/globalmount", p.Name)
+			if err := ns.mount(mountPath, p.Name, p.Spec.CSI.VolumeAttributes); err != nil {
+				glog.Warningf("remount damaged volume %q to path %q failed: %v\n", p.Name, mountPath, err)
+				return
+			}
+			glog.Infof("remount damaged volume %q to path %q succeed.", p.Name, mountPath)
+
+			// bind globalmount to pods
+			for _, pod := range p.pods {
+				bindTargetPath := fmt.Sprintf("/var/lib/kubelet/pods/%s/volumes/kubernetes.io~csi/%s/mount", pod.UID, p.Name)
```

Comment:
the root-dir prefer to be configurable, --root-dir may be not "/var/lib/kubelet", in our k8s, it's "/data/kubernetes/var/lib/kubelet".

Reply:
done.